contracts/src/lib/LibGateway.sol

```diff
@@ -386,20 +400,149 @@ library LibGateway {
             supplySource = SupplySourceHelper.native();
         }
 
+        // If the crossnet destination is NOT the current network (network where the gateway is running),
+        // we add it to the postbox for further propagation.
+        // Even if we send for propagation, the execution of every message
+        // should increase the appliedNonce to allow the execution of the next message
+        // of the batch (this is way we have this after the nonce logic).
+        if (!crossMsg.to.subnetId.equals(s.networkName)) {
+            bytes32 cid = crossMsg.toHash();
+            s.postbox[cid] = crossMsg;
+            return;
+        }
+
+        // execute the message and get the receipt.
+        (bool success, bytes memory ret) = executeCrossMsg(crossMsg, supplySource);
+        if (success) {
+            sendReceipt(crossMsg, OutcomeType.Ok, ret);
+        } else {
+            sendReceipt(crossMsg, OutcomeType.ActorErr, ret);
+        }
+    }
+
+    /// @dev Execute the cross message using low level `call` method. This way ipc will
+    ///      catch contract revert messages as well. We need this because in `CrossMsgHelper.execute`
+    ///      there are `require` and `revert` calls, without reflexive call, the execution will
+    ///      revert and block the checkpoint submission process.
+    function executeCrossMsg(IpcEnvelope memory crossMsg, SupplySource memory supplySource) internal returns (bool success, bytes memory result) {
+        (success, result) = address(CrossMsgHelper).delegatecall(   // solhint-disable-line avoid-low-level-calls
+            abi.encodeWithSelector(CrossMsgHelper.execute.selector, crossMsg, supplySource)
+        );
+
+        if (success) {
+            return abi.decode(result, (bool, bytes));
+        }
+
+        return (success, result);
+    }
+
+    /// @notice Sends a receipt from the execution of a cross-message.
+    /// Only `Call` messages trigger a receipt. Transfer messages should be directly
+    /// handled by the peer client to return the funds to the from address in the
+    /// failing network.
+    /// (we could optionally trigger a receipt from `Transfer`s to, but without
+    /// multi-level execution it would be adding unnecessary overhead).
+    function sendReceipt(IpcEnvelope memory crossMsg, OutcomeType outcomeType, bytes memory ret) internal {
+        if (crossMsg.isEmpty()) {
+            revert CannotCreateIpcReceipt();
+        }
```

Comment:
My take on this is a sanity check. Ideally `crossMsg` should not have been empty. If this happens that means validators have agreed to submit empty cross message. It will block checkpoint from being committed. Personally it's hard to tell if this is good or bad because if this happens, something is really wrong among the validators. But on the other hand, if validators all agree on empty cross msg, so be it, we can just return for the submission to go on.